README.md

# e-Kasa Cloud Client PHP

HTTP klient v jazyku PHP pre [e-Kasa cloudové riešenie](https://github.com/ninedigit/ekasa-cloud) spoločnosti [Nine Digit, s.r.o.](https://ekasa.ninedigit.sk/).

Knižnica je kompatibilná s PHP verzie 7.4 a vyššie.

# Vývoj

## Inštalácia Composer-a

```bash
php -r "copy('https://getcomposer.org/installer', 'composer-setup.php');"
php -r "if (hash_file('sha384', 'composer-setup.php') === '55ce33d7678c5a611085589f1f3ddf8b3c52d662cd01d4ba75c0ee0459970c2200a51f492d557530c71c15d8dba01eae') { echo 'Installer verified'; } else { echo 'Installer corrupt'; unlink('composer-setup.php'); } echo PHP_EOL;"
php composer-setup.php
php -r "unlink('composer-setup.php');"
```

Inštalácia závislostí knižnice: `php composer.phar install`.

> Po vytvorení nových tried (v separátnych súboroch) je nutné vyvolať `php composer.phar dump-autoload`.

# Testovanie

Pre vykonanie integračných testov je nutné vytvoriť súbor `settings.json` v `tests\integration` vo formáte **JSON** so štruktúrou zhodnou s triedou `ApiClientOptions` a teda:

```json
{
    "publicKey": "PUBLIC_KEY",
    "privateKey": "PRIVATE_KEY",
    "tenantId": "TENANT_ID",
    "url": "https://ekasa-cloud.ninedigit.sk/api",
    "proxyUrl": null,
    "cashRegisterCode": "ORP_CODE"
}
```

Testy je možné spustiť príkazom
`./vendor/bin/phpunit --verbose tests`.

Pre spustenie konkrétneho testu je nutné uviesť prepínač `--filter` s názvom testovaciej metódy a teda `./vendor/bin/phpunit --verbose tests --filter testRegisterReceiptUsingPosPrinterWithNonEmptyOptions`.

## Overenie kompatibility PHP

`/vendor/bin/phpcs --standard=PHPCompatibility --extensions=php --runtime-set testVersion 7.4- ./src`

# Inštalácia

Upravte súbor `composer.json` a pridajte nový repozitár:

```
{
  "repositories": [
    {
      "type": "vcs",
      "url": "https://github.com/ninedigit/ekasa-cloud-clientphp"
    }
  ]
}
```

Následne nainštalujte balík príkazom `composer require ninedigit/ekasa-cloud-clientphp`. Pre inštaláciu špecifickej verzie bude príkaz vyzerať nasledovne `composer require ninedigit/ekasa-cloud-clientphp:1.0.0`.

> Pre viac informácií navštívte https://getcomposer.org/doc/05-repositories.md#using-private-repositories.

# Použitie

> Príklady sú dostupné v adresári [examples](https://github.com/ninedigit/ekasa-cloud-clientphp/tree/master/examples).

## Zaokrúhľovanie

Výsledky všetkých matematických operácií musia byť pred priradením do príslušných premenných zaokrúhlené a to buď na 6 desatinných miest pre jednotkové ceny (napr. vlastnosť `ReceiptRegistrationItemDto.unitPrice`) alebo na dve desatinné miesta pre ostatné ceny, prípadne množstvá (napr. vlastnosť `QuantityDto.amount` alebo `ReceiptRegistrationItemDto.price`).

## Príklad registrácie požiadavky
```php
// Nastavenia tlače
$receiptPrinter = new PosReceiptPrinterDto();

// Položka
$receiptItem = new ReceiptRegistrationItemDto(
  ReceiptItemType::POSITIVE,
  "Coca Cola 0.25l",
  1.29,
  20.00,
  new QuantityDto(2, "ks"), 2.58);

// Platba
$receiptPayment = new ReceiptRegistrationPaymentDto(2.58, "Hotovosť");

// Požiadavka registrácie
$receiptRegistrationRequest = new CreateReceiptRegistrationRequestDto(
  ReceiptType::CASH_REGISTER,
  "88812345678900001",
  "e52ff4d1-f2ed-4493-9e9a-a73739b1ba23",
  [$receiptItem],
  [$receiptPayment]);

// Registrácia
$createReceiptRegistration = new CreateReceiptRegistrationDto(
  $receiptPrinter, $receiptRegistrationRequest, 10000);
$receiptRegistration = $client->registerReceipt($createReceiptRegistration);

if ($receiptRegistration->isCompletedSuccessfully()) {
  // ...
}
```

CHANGELOG.md
# Changelog

Všetky významné zmeny v projekte budú uvedené v tomto dokumente.

Formát dokumentu vychádza z [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
a dodržiava [sémantické verzionovanie](https://semver.org/spec/v2.0.0.html).

## [0.9.16] - 2022-12-22

### Pridané

- nová vlastnosť `userMeta` pre triedy `CustomerDto`, `CustomerCreditBalanceTransactionDto` a `CustomerCardDto`

## [0.9.15] - 2022-11-18

- Chýbajúce informácie o vydaných verziách do CHANGELOG.md

### Zmenené

- Menný priestor pre testy

## [0.9.14] - 2022-11-16

### Pridané

- Informácie o zaokrúhľovaní do README.MD
- Pole `cashRegisterCode` do konfiguračného súboru integračných testov
- Inforácie o inštalácií Composer-a
- Podpora prijatia pola hodnôt ako argumentu pre metódu `load` triedy `NineDigit\eKasa\Cloud\Client\ApiClientOptions`

### Zmenené

- Názov konfiguračného súboru pre integračné testy z `ApiClientOptions.json` na `settings.json`
- Reflektované zmeny do README.MD


## [0.9.13] - 2022-10-20

- Opravená chyba, kedy prázdny objekt bol serializovaný ako prázdne pole (`[]`).
- Opravený príklad registrácie dokladu

### Pridané

- Integračné testy pre triedu `NineDigit\eKasa\Cloud\Client\ApiClient`
- Unit testy pre JSON serializér

## [0.9.12] - 2022-10-19

- Odstránené použitie jazykových súčastí nekompatibilných s verziu PHP 7.4
- Opravená chyba (Ticket #6696) kedy vyvolanie metódy `registerReceipt` triedy `NineDigit\eKasa\Cloud\Client\ApiClient` mohlo zlyhať

### Pridané

- Unit testy pre triedu `NineDigit\eKasa\Cloud\Client\ApiRequestBuilder`

## [0.9.11] - 2022-08-03

- Odstránené použitie menných parametrov, ktoré spôsobovalo nekompatibilitu s PHP 7.4

## [0.9.10] - 2022-08-02

### Pridané

- Opravená chyba, kedy sa neprenášal parameter `cardSerialNumbers` do dopytu zákazníkov
- Vylepšenie integračného testu `testGetCustomersWithFilter` triedy `NineDigit\eKasa\Cloud\Client\ApiClientTest`
- Unit testy pre triedu `NineDigit\eKasa\Cloud\Client\CustomerFilterDto`
- Pridaná možnosť overenia kompatibility s verziou PHP 7.4
- Oprava príkladov na vyčítanie zákanika/ov

### Zmenené

- Premenovaná trieda `NineDigit\eKasa\Cloud\Client\ApiClientTest` na `NineDigit\eKasa\Cloud\Client\ApiClientIntegrationTest`

## [0.9.9] - 2022-08-02

### Pridané

- Integračné testy
- Metóda `load` do triedy `ApiClientOptions` na vyčítanie konfigurácie zo súboru

## [0.9.8] - 2022-08-02

### Pridané

- Príklady volania na získanie zákazníckych účtov
- Metóda `findCustomer` do triedy `ApiClient` na vyhľadanie zákaníckeho účtu

## [0.9.7] - 2022-06-26

### Pridané

- Metódy do triedy `ApiClient` na získanie zákazníkov
- Pridané vlastnosti `serializer` a `requestMessageSigner` pre lepšiu testovateľnosť triedy `ApiClient`
- Pridaná metóda `isSuccessStatusCode` a `ensureSuccessStatusCode` do triedy `ApiResponseMessage` na overenie/zabezpečenie kladnej odpovede zo servera
- Pridaný parameter `url` do `ApiClientOptions`
- Testy triedy `ApiClient`

### Zmenené

- Metóda `withDefaultHeaders` triedy `ApiRequestBuilder` bola premenovaná na `withHeaders` a preberá buď asociatívne pole hlavičiek alebo volateľnú metódu na nastavenie hlavičiek
- Metóda `setDefault` triedy `ApiRequestHeadersBuilder` bola premenovaná na `set`
- Interne upravená trieda `ApiClient` pre lepšiu testovateľnosť

## [0.9.6] - 2021-11-09

### Pridané

- Zoznam vrátených chybových kódov z API - `ApiErrorCode`
- Pridaná vlastnosť `$code` do `ProblemDetails` nesúca chybový kód odpovede z API
- Zoznam dostupných prostredí - `CloudEnvironment`
- Test deserializácie `ProblemDetails` objektu

### Zmenené

- Konštanta `DEFAULT_URL` a vlastnosť `$tenantKey` triedy `ApiClientOptions` boli označené ako `@deprecated` a budú odstránené v nasledujúcej verzií


examples/index.php
<?php

namespace NineDigit\eKasa\Cloud\Client\Examples;

use \Error;
use NineDigit\eKasa\Cloud\Client\ApiClient;
use NineDigit\eKasa\Cloud\Client\ApiClientOptions;
use NineDigit\eKasa\Cloud\Client\Models\QuantityDto;
use NineDigit\eKasa\Cloud\Client\Models\Registrations\Receipts;
use NineDigit\eKasa\Cloud\Client\Models\Registrations\Receipts\PosReceiptPrinterOptions;
use NineDigit\eKasa\Cloud\Client\Models\Registrations\Receipts\PosReceiptPrinterDto;
use NineDigit\eKasa\Cloud\Client\Models\Registrations\Receipts\PdfReceiptPrinterDto;
use NineDigit\eKasa\Cloud\Client\Models\Registrations\Receipts\EmailReceiptPrinterOptions;
use NineDigit\eKasa\Cloud\Client\Models\Registrations\Receipts\EmailReceiptPrinterDto;
use NineDigit\eKasa\Cloud\Client\Models\Registrations\Receipts\CreateReceiptRegistrationDto;
use NineDigit\eKasa\Cloud\Client\Models\Registrations\Receipts\ReceiptItemType;
use NineDigit\eKasa\Cloud\Client\Models\Registrations\Receipts\ReceiptRegistrationItemDto;
use NineDigit\eKasa\Cloud\Client\Models\Registrations\Receipts\ReceiptRegistrationPaymentDto;
use NineDigit\eKasa\Cloud\Client\Models\Registrations\Receipts\ReceiptType;
use NineDigit\eKasa\Cloud\Client\Models\Registrations\RegistrationState;

require '../vendor/autoload.php';

//////////////////////////////////////////////////

// Verejný kľúč pre prístup k API
$publicKey = "public_key";
// Súkromný kľúč pre prístup k API
$privateKey = "private_key";
// Identifikátor skupiny koncových zariadení
$tenantId = "tenant_id";

// Inštanciácia API klienta
$clientOptions = new ApiClientOptions($publicKey, $privateKey, $tenantId);
$client = new ApiClient($clientOptions);

// Kód on-line registračnej pokladne (ORP kód), ktorá požiadavku spracuje
$cashRegisterCode = "88812345678900001";
/*
 * Identifikátor požiadavky, priradený nadradenou aplikáciou.
 * Spolu s ORP kódom tvoria unikátny identifikátor požidavky naprieč systémom.
 */
$externalId = "e52ff4d1-f2ed-4493-9e9a-a73739b1ba23";

/**
 * Konfigurácia tlačiarne, na ktorej bude doklad spracovaný.
 * K dispozícií sú nasledovné:
 */

  // 1. Tlačiareň papierových dokladov
  $posPrinterOptions = new PosReceiptPrinterOptions();
  $posPrinterOptions->openDrawer = true;
  $receiptPrinter = new PosReceiptPrinterDto($posPrinterOptions);

  // 2. Tlačiareň vyhotovujúca PDF súbory
  $receiptPrinter = new PdfReceiptPrinterDto($pdfPrinterOptions);

  // 3. Tlačiareň vyhotovujúca e-maily
  $emailPrinterOptions = new EmailReceiptPrinterOptions("mail@example.com");
  $emailPrinterOptions->subject = "Váš elektronický doklad";
  $receiptPrinter = new EmailReceiptPrinterDto($emailPrinterOptions);

// Položka dokladu
$item = new ReceiptRegistrationItemDto(
  ReceiptItemType::POSITIVE,  // Kladný typ položky
  "Coca Cola 0.25l",          // Názov
  1.29,                       // Jednotková cena
  20.00,                      // Daňová hladina
  new QuantityDto(2, "ks"),   // Množstvo
  2.58                        // Cena
);

// Platba
$payment = new ReceiptRegistrationPaymentDto(2.58, "Hotovosť");

$receiptRegistrationRequest = new Receipts\CreateReceiptRegistrationRequestDto(
  ReceiptType::CASH_REGISTER, // Typ dokladu
  $cashRegisterCode,          // ORP kód
  $externalId,                // Externý identifikátor
  [$item],                    // Zoznam položiek
  [$payment]                  // Zoznam platidiel
);

// Voliteľná hlavička a pätička dokladu
$receiptRegistrationRequest->headerText = "www.ninedigit.sk";
$receiptRegistrationRequest->footerText = "Ďakujeme za váš nákup.";

/*
 * Maximálna platnosť požiadavky registrácie dokladu v milisekundách.
 * Ak počas tohto intervalu nedôjde k akceptácií požiadavky dokladu
 * registračnou pokladňou, požiadavka bude zrušená.
 */
$validityTimeSpan = 10000; // 10 sekúnd

// Kontext požidavky registrácie dokladu
$createReceiptRegistration = new CreateReceiptRegistrationDto(
  $receiptPrinter, $receiptRegistrationRequest, $validityTimeSpan);

// Zaslanie požidavky a obdržanie výsledku
$receiptRegistration = $client->registerReceipt($createReceiptRegistration);

// Doklad bol úspešne spracovaný v režíme ON-LINE
if ($receiptRegistration->state === RegistrationState::PROCESSED) {
  
}
// Doklad bol úspešne spracovaný v režíme OFF-LINE
else if ($receiptRegistration->state === RegistrationState::PROCESSED_OFFLINE) {
  
}
// Spracovanie dokladu zlyhalo zo strany e-Kasa systému
else if ($receiptRegistration->state === RegistrationState::PROCESS_FAILED) {
  $errorMessage = $receiptRegistration->request->eKasaError->message;
  $errorCode = $receiptRegistration->request->eKasaError->code;
  throw new Error(`Spracovanie požiadavky bolo zamietnuté: ${errorMessage} [${errorCode}]`);
}
// Spracovanie dokladu zlyhalo
else if ($receiptRegistration->state === RegistrationState::FAILED) {
  $errorMessage = $receiptRegistration->error->message;
  throw new Error(`Spracovanie požiadavky zlyhalo: ${errorMessage}.`);
}
else {
  // Odpoveď môže byť v jednom zo stavov: EXPIRED, TIMED_OUT, CANCELED alebo FAILED.
}

?>